File: javascript/quick_sort.js
function quickSort(arr) {
  sort(arr, 0, arr.length - 1);
}

function sort(arr, start, end) {
  if (start >= end) {
    return;
  }

  const pivot = partition(arr, start, end);

  sort(arr, start, pivot - 1);
  sort(arr, pivot + 1, end);
}

function partition(arr, start, end) {
  const pivot = arr[end];
  let index = start - 1;

  for (let i = start; i < end; i++) {
    if (arr[i] <= pivot) {
      index++;

      const temp = arr[i];
      arr[i] = arr[index];
      arr[index] = temp;
    }
  }

  index++;
  arr[end] = arr[index];
  arr[index] = pivot;

  return index;
}

const arr = [9, 1, 5, 13, 14, 19, 2, 8, 7, 3, 0, 21, 8, 75, 10];
quickSort(arr)
console.log(arr);

File: java/kata/src/main/java/br/com/ari/algo/algorithms/QuickSort.java
package br.com.ari.algo.algorithms;

public class QuickSort {

    private QuickSort() {}

    public static void sort(int[] arr) {
        sort(arr, 0, arr.length - 1);
    }

    private static void sort(int[] arr, int low, int high) {
        if (low >= high) {
            return;
        }

        final int pivot = partition(arr, low, high);

        sort(arr, low, pivot - 1);
        sort(arr, pivot + 1, high);
    }

    private static int partition(int[] arr, int low, int high) {
        final int pivot = arr[high];
        int index = low - 1;

        for (int i = low; i < high; i++) {
            if (arr[i] <= pivot) {
                swap(arr, i, ++index);
            }
        }

        swap(arr, high, ++index);

        return index;
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

}
